core/slot_chain.go

```diff
@@ -0,0 +1,184 @@
+package core
+
+import (
+	"log"
+	"sync"
+)
+
+// _defaultSlotChain is a default slot chain built by framework
+// _defaultSlotChain is global unique chain
+var _defaultSlotChain = buildDefaultSlotChain()
+
+// StatPrepareSlot is responsible for some preparation before statistic
+// For example: init structure and so on
+type StatPrepareSlot interface {
+	// Prepare function do some initialization
+	// Such as: init statistic structure、node and etc
+	// The result of preparing would store in Context
+	// All StatPrepareSlots execute in sequence
+	// Prepare function should not throw panic.
+	Prepare(ctx *Context)
+}
+
+// RuleCheckSlot is rule based checking strategy
+// All checking rule must implement this interface.
+type RuleCheckSlot interface {
+	// Check function do some validation
+	// It can break off the slot pipeline
+	// Each RuleCheckSlot will return check result
+	// The upper logic will control pipeline according to SlotResult.
+	Check(ctx *Context) *RuleCheckResult
+}
+
+// StatSlot is responsible for counting all custom biz metrics.
+type StatSlot interface {
+	// OnEntryPass function will be invoked when StatPrepareSlots and RuleCheckSlots execute pass
+	// StatSlots will do some statistic logic, such as QPS、log、etc
+	OnEntryPassed(ctx *Context)
+	// OnEntryBlocked function will be invoked when StatPrepareSlots and RuleCheckSlots execute fail
+	// StatSlots will do some statistic logic, such as QPS、log、etc
+	OnEntryBlocked(ctx *Context)
```

Comment:
Maybe we need an additional parameter to indicate the type of blocking event (like BlockException)?

Reply:
I have refined here and add  a blocked event as parameter.